### PR TITLE
Pass F::*FuncOptions instead of torch::nn::*Options to functionals, and make F::*FuncOptions a different class when necessary

### PR DESCRIPTION
**Pull request by yf225**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29364 Pass F::*FuncOptions instead of torch::nn::*Options to functionals, and make F::*FuncOptions a different class when necessary **
* #29360 [C++ API] Add functional overloads for fold, linear, loss, normalization, padding
* #29359 [C++ API] Add functional overloads for pixelshuffle, pooling, upsampling, vision
* #29358 [C++ API] Add functional overloads for activation, batchnorm, distance, embedding

Currently, we use `torch::nn::*Options` both as module options and functional options. However, this makes it very hard to manage the parameters in `torch::nn::*Options`, because a module's constructor can take a different set of arguments than the module's equivalent functional (e.g. `torch.nn.BatchNorm1d` takes `num_features, eps=1e-5, momentum=0.1, affine=True,
track_running_stats=True`, while `F::batch_norm` takes `running_mean, running_var, weight=None, bias=None, training=False, momentum=0.1, eps=1e-5`).

This PR resolves the above problem by making `F::*FuncOptions` a different class from `torch::nn::*Options` when necessary (i.e. when a module's constructor takes a different set of arguments than the module's equivalent functional). In the rest of the cases where the module constructor takes the same set of arguments as the module's equivalent functional, `F::*FuncOptions` is an alias of `torch::nn::*Options`.

Also as part of this PR, we change all functional options to pass-by-value, to make the semantics consistent across all functionals.

Differential Revision: [D18376977](https://our.internmc.facebook.com/intern/diff/D18376977)

### CODE REVIEW

**Comment by kostmo** [comment]
## CircleCI build failures summary
As of commit f7d9dec:
* 0/1 recognized as flaky
* 1/1 failures introduced in this PR



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=f7d9dec925558de1b3aaa746301d530ae29f0fbc).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

File: torch/csrc/api/include/torch/nn/options/vision.h [hunk 1]
```diff
@@ -6,9 +6,10 @@
 
 namespace torch {
 namespace nn {
+namespace functional {
 
 /// Options for a Grid Sample module.
```
Comment (pbelevich) [inline torch/csrc/api/include/torch/nn/options/vision.h]:
```suggestion
/// Options for a Grid Sample functional.
```

**Review by pbelevich — APPROVED** [review]

File: torch/csrc/api/include/torch/nn/options/vision.h (hunk 1 above)
Comment (yf225) [reply torch/csrc/api/include/torch/nn/options/vision.h]: Fixed

File: torch/csrc/api/include/torch/nn/functional/loss.h [hunk 2]
```diff
@@ -139,7 +139,7 @@ inline Tensor binary_cross_entropy(
 inline Tensor binary_cross_entropy(
     const Tensor& input,
     const Tensor& target,
-    const BCELossOptions& options = {}) {
+    BCELossFuncOptions options = {}) {
```
[the remark this answers is not in the document]
Comment (yf225) [reply torch/csrc/api/include/torch/nn/functional/loss.h]: Passing a `Tensor` object by value actually just increments its `TensorImpl` intrusive_ptr counter, and doesn't actually deep-copy the tensor data storage, so I think it's okay to consider it as a light-weight object.

File: torch/csrc/api/include/torch/nn/functional/loss.h [hunk 3]
```diff
@@ -195,7 +195,7 @@ inline Tensor multi_margin_loss(
 inline Tensor multi_margin_loss(
     const Tensor& input,
     const Tensor& target,
-    const MultiMarginLossOptions& options = {}) {
+    MultiMarginLossFuncOptions options = {}) {
```
[the remark this answers is not in the document]
Comment (yf225) [reply torch/csrc/api/include/torch/nn/functional/loss.h]: replied in an earlier comment

File: torch/csrc/api/include/torch/nn/functional/loss.h [hunk 4]
```diff
@@ -344,7 +344,7 @@ inline Tensor multilabel_soft_margin_loss(
 inline Tensor multilabel_soft_margin_loss(
     const Tensor& input,
     const Tensor& target,
-    const MultiLabelSoftMarginLossOptions& options = {}) {
+    MultiLabelSoftMarginLossFuncOptions options = {}) {
```
[the remark this answers is not in the document]
Comment (yf225) [reply torch/csrc/api/include/torch/nn/functional/loss.h]: replied in an earlier comment
